File: backend/app/models/inventory.py
import uuid
from datetime import date, datetime
from typing import Optional
from sqlalchemy import (
    Boolean,
    Date,
    DateTime,
    ForeignKey,
    Integer,
    Numeric,
    String,
    UniqueConstraint,
    Enum as SQLEnum,
)
from sqlalchemy.dialects.postgresql import UUID
from sqlalchemy.orm import Mapped, mapped_column, relationship
import enum
from app.database import Base
from app.models._mixins import uuid_pk, created_at_col, updated_at_col

# Forward-declare enums used on Inventory (defined in copilot.py to avoid circular imports)
_INVENTORY_TYPE_ENUM = "inventory_type_enum"
_SOURCING_STRATEGY_ENUM = "sourcing_strategy_enum"


class Category(Base):
    __tablename__ = "categories"

    id: Mapped[uuid_pk]
    parent_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("categories.id"), nullable=True
    )
    catg_code: Mapped[str] = mapped_column(String(50), nullable=False)
    cag_catg_code: Mapped[Optional[str]] = mapped_column(String(50), nullable=True)
    description: Mapped[str] = mapped_column(String(255), nullable=False)
    store_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("stores.id", ondelete="CASCADE"), nullable=False
    )
    created_at: Mapped[created_at_col]
    updated_at: Mapped[updated_at_col]

    # Relationships
    parent = relationship("Category", remote_side="Category.id", lazy="raise")
    store = relationship("Store", back_populates="categories", lazy="raise")
    skus = relationship("SKU", back_populates="category", lazy="raise")
    promotions = relationship("Promotion", back_populates="category", lazy="raise")

    def __repr__(self) -> str:
        return f"<Category {self.catg_code}: {self.description}>"


class Brand(Base):
    __tablename__ = "brands"

    id: Mapped[uuid_pk]
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    category_type: Mapped[Optional[str]] = mapped_column(String(100), nullable=True)
    created_at: Mapped[created_at_col]

    # Relationships
    skus = relationship("SKU", back_populates="brand", lazy="raise")

    def __repr__(self) -> str:
        return f"<Brand {self.name}>"


class SKU(Base):
    __tablename__ = "skus"

    id: Mapped[uuid_pk]
    sku_code: Mapped[str] = mapped_column(
        String(16), unique=True, index=True, nullable=False
    )
    description: Mapped[str] = mapped_column(String(60), nullable=False)
    long_description: Mapped[Optional[str]] = mapped_column(String(1000), nullable=True)
    cost_price: Mapped[Optional[float]] = mapped_column(Numeric(20, 2), nullable=True)
    category_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("categories.id"), nullable=True
    )
    brand_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("brands.id"), nullable=True
    )
    tax_code: Mapped[str] = mapped_column(String(1), nullable=False, default="G")
    gender: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    age_group: Mapped[Optional[str]] = mapped_column(String(20), nullable=True)
    is_unique_piece: Mapped[bool] = mapped_column(
        Boolean, default=False, nullable=False
    )
    use_stock: Mapped[bool] = mapped_column(Boolean, default=True, nullable=False)
    block_sales: Mapped[bool] = mapped_column(Boolean, default=False, nullable=False)
    store_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("stores.id", ondelete="CASCADE"), nullable=False
    )
    created_at: Mapped[created_at_col]
    updated_at: Mapped[updated_at_col]

    # Relationships
    category = relationship("Category", back_populates="skus", lazy="raise")
    brand = relationship("Brand", back_populates="skus", lazy="raise")
    store = relationship("Store", back_populates="skus", lazy="raise")
    plus = relationship("PLU", back_populates="sku", lazy="raise")
    prices = relationship("Price", back_populates="sku", lazy="raise")
    promotions = relationship("Promotion", back_populates="sku", lazy="raise")
    inventories = relationship("Inventory", back_populates="sku", lazy="raise")
    order_items = relationship("OrderItem", back_populates="sku", lazy="raise")

    def __repr__(self) -> str:
        return f"<SKU {self.sku_code}: {self.description}>"


class PLU(Base):
    __tablename__ = "plus"

    id: Mapped[uuid_pk]
    plu_code: Mapped[str] = mapped_column(
        String(20), unique=True, index=True, nullable=False
    )
    sku_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("skus.id", ondelete="CASCADE"), nullable=False
    )
    created_at: Mapped[created_at_col]

    # Relationships
    sku = relationship("SKU", back_populates="plus", lazy="raise")

    def __repr__(self) -> str:
        return f"<PLU {self.plu_code}>"


class Price(Base):
    __tablename__ = "prices"

    id: Mapped[uuid_pk]
    sku_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("skus.id", ondelete="CASCADE"), nullable=False
    )
    store_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("stores.id", ondelete="CASCADE"), nullable=True
    )
    price_incl_tax: Mapped[float] = mapped_column(Numeric(20, 2), nullable=False)
    price_excl_tax: Mapped[float] = mapped_column(Numeric(20, 2), nullable=False)
    price_unit: Mapped[int] = mapped_column(Integer, default=1, nullable=False)
    valid_from: Mapped[date] = mapped_column(Date, nullable=False)
    valid_to: Mapped[date] = mapped_column(Date, nullable=False)
    created_at: Mapped[created_at_col]
    updated_at: Mapped[updated_at_col]

    # Relationships
    sku = relationship("SKU", back_populates="prices", lazy="raise")
    store = relationship("Store", back_populates="prices", lazy="raise")

    def __repr__(self) -> str:
        return f"<Price SKU={self.sku_id} ${self.price_incl_tax}>"


class Promotion(Base):
    __tablename__ = "promotions"

    id: Mapped[uuid_pk]
    disc_id: Mapped[str] = mapped_column(String(20), nullable=False)
    sku_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("skus.id", ondelete="CASCADE"), nullable=True
    )
    category_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True),
        ForeignKey("categories.id", ondelete="CASCADE"),
        nullable=True,
    )
    line_type: Mapped[str] = mapped_column(String(20), nullable=False)
    disc_method: Mapped[str] = mapped_column(String(20), nullable=False)
    disc_value: Mapped[float] = mapped_column(Numeric(11, 2), nullable=False)
    line_group: Mapped[Optional[str]] = mapped_column(String(1), nullable=True)
    created_at: Mapped[created_at_col]
    updated_at: Mapped[updated_at_col]

    # Relationships
    sku = relationship("SKU", back_populates="promotions", lazy="raise")
    category = relationship("Category", back_populates="promotions", lazy="raise")

    def __repr__(self) -> str:
        return f"<Promotion {self.disc_id}>"


class InventoryLocationState(str, enum.Enum):
    STORE = "STORE"
    TRANSIT = "TRANSIT"
    WORKSHOP = "WORKSHOP"
    DECOR = "DECOR"


class Inventory(Base):
    __tablename__ = "inventories"
    __table_args__ = (
        UniqueConstraint("store_id", "sku_id", name="uq_inventory_store_sku"),
    )

    id: Mapped[uuid_pk]
    location_status: Mapped[InventoryLocationState] = mapped_column(
        SQLEnum(InventoryLocationState, name="inventory_location_enum"), 
        default=InventoryLocationState.STORE, 
        nullable=False
    )
    sku_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("skus.id", ondelete="CASCADE"), nullable=False
    )
    store_id: Mapped[uuid.UUID] = mapped_column(
        UUID(as_uuid=True), ForeignKey("stores.id", ondelete="CASCADE"), nullable=False
    )
    qty_on_hand: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    reorder_level: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    reorder_qty: Mapped[int] = mapped_column(Integer, default=0, nullable=False)
    serial_number: Mapped[Optional[str]] = mapped_column(String(255), nullable=True)
    last_updated: Mapped[datetime] = mapped_column(DateTime, nullable=False)
    # Copilot fields
    inventory_type: Mapped[str] = mapped_column(
        SQLEnum("purchased", "material", "finished", name=_INVENTORY_TYPE_ENUM, create_type=False),
        nullable=False, default="purchased",
    )
    sourcing_strategy: Mapped[str] = mapped_column(
        SQLEnum("supplier_premade", "manufactured_standard", "manufactured_custom",
                name=_SOURCING_STRATEGY_ENUM, create_type=False),
        nullable=False, default="supplier_premade",
    )
    primary_supplier_id: Mapped[Optional[uuid.UUID]] = mapped_column(
        UUID(as_uuid=True), ForeignKey("suppliers.id", ondelete="SET NULL"), nullable=True
    )
    created_at: Mapped[created_at_col]
    updated_at: Mapped[updated_at_col]

    # Relationships
    sku = relationship("SKU", back_populates="inventories", lazy="raise")
    store = relationship("Store", back_populates="inventories", lazy="raise")

    def __repr__(self) -> str:
        return f"<Inventory SKU={self.sku_id} qty={self.qty_on_hand}>"

File: apps/ios/retailmanagement/Models/Inventory.swift
//
//  Inventory.swift
//  retailmanagement
//

import Foundation

nonisolated struct Category: Codable, Identifiable, Sendable {
    let id: String
    let catgCode: String
    let cagCatgCode: String?
    let description: String
    let parentId: String?
    let storeId: String
    let createdAt: String?
    let updatedAt: String?
}

nonisolated struct Brand: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let categoryType: String?
    let createdAt: String?
}

nonisolated struct SKU: Codable, Identifiable, Hashable, Sendable {
    static func == (lhs: SKU, rhs: SKU) -> Bool { lhs.id == rhs.id }
    func hash(into hasher: inout Hasher) { hasher.combine(id) }

    let id: String
    let skuCode: String
    let description: String
    let longDescription: String?
    let costPrice: Double?
    let categoryId: String?
    let brandId: String?
    let taxCode: String
    let gender: String?
    let ageGroup: String?
    let isUniquePiece: Bool
    let useStock: Bool
    let blockSales: Bool
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let supplierSkuCode: String?
    let storeId: String
    let createdAt: String?
    let updatedAt: String?

    var displayPrice: String {
        if let cost = costPrice {
            return String(format: "$%.2f", cost)
        }
        return "N/A"
    }
}

nonisolated enum InventoryType: String, Codable, Sendable {
    case purchased
    case material
    case finished

    var displayName: String {
        switch self {
        case .purchased: return "Purchased"
        case .material: return "Material"
        case .finished: return "Finished"
        }
    }
}

nonisolated enum SourcingStrategy: String, Codable, Sendable {
    case supplierPremade = "supplier_premade"
    case manufacturedStandard = "manufactured_standard"
    case manufacturedCustom = "manufactured_custom"

    var displayName: String {
        switch self {
        case .supplierPremade: return "Supplier pre-made"
        case .manufacturedStandard: return "Manufactured standard"
        case .manufacturedCustom: return "Manufactured custom"
        }
    }
}

nonisolated struct InventoryItem: Codable, Identifiable, Sendable {
    let id: String
    let skuId: String
    let storeId: String
    let qtyOnHand: Int
    let reorderLevel: Int
    let reorderQty: Int
    let serialNumber: String?
    let lastUpdated: String
    let createdAt: String?
    let updatedAt: String?

    var isLowStock: Bool {
        qtyOnHand <= reorderLevel
    }
}

nonisolated struct Price: Codable, Identifiable, Sendable {
    let id: String
    let skuId: String
    let storeId: String?
    let priceInclTax: Double
    let priceExclTax: Double
    let priceUnit: Int
    let validFrom: String
    let validTo: String
    let createdAt: String?
    let updatedAt: String?

    var formattedPrice: String {
        String(format: "$%.2f", priceInclTax)
    }
}

nonisolated struct Promotion: Codable, Identifiable, Sendable {
    let id: String
    let discId: String
    let skuId: String?
    let categoryId: String?
    let lineType: String
    let discMethod: String
    let discValue: Double
    let lineGroup: String?
    let createdAt: String?
    let updatedAt: String?
}

nonisolated enum JSONValue: Codable, Sendable {
    case string(String)
    case number(Double)
    case bool(Bool)
    case object([String: JSONValue])
    case array([JSONValue])
    case null

    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        if container.decodeNil() {
            self = .null
        } else if let value = try? container.decode(Bool.self) {
            self = .bool(value)
        } else if let value = try? container.decode(Double.self) {
            self = .number(value)
        } else if let value = try? container.decode(String.self) {
            self = .string(value)
        } else if let value = try? container.decode([String: JSONValue].self) {
            self = .object(value)
        } else if let value = try? container.decode([JSONValue].self) {
            self = .array(value)
        } else {
            throw DecodingError.dataCorruptedError(in: container, debugDescription: "Unsupported JSON value")
        }
    }

    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        switch self {
        case .string(let value):
            try container.encode(value)
        case .number(let value):
            try container.encode(value)
        case .bool(let value):
            try container.encode(value)
        case .object(let value):
            try container.encode(value)
        case .array(let value):
            try container.encode(value)
        case .null:
            try container.encodeNil()
        }
    }
}

nonisolated enum RecommendationType: String, Codable, Sendable {
    case reorder
    case priceChange = "price_change"
    case stockAnomaly = "stock_anomaly"

    var displayName: String {
        switch self {
        case .reorder: return "Reorder"
        case .priceChange: return "Price Review"
        case .stockAnomaly: return "Stock Anomaly"
        }
    }
}

nonisolated enum RecommendationStatus: String, Codable, Sendable {
    case pending
    case approved
    case rejected
    case applied
    case expired
    case queued
    case unavailable

    var displayName: String { rawValue.capitalized }
}

nonisolated struct RecommendationOutcome: Codable, Identifiable, Sendable {
    let recommendationId: String
    let skuId: String?
    let title: String
    let type: RecommendationType
    let status: RecommendationStatus
    let updatedAt: String?

    var id: String { recommendationId }
}

nonisolated struct ManagerSummary: Codable, Sendable {
    let storeId: String
    let analysisStatus: String
    let lastGeneratedAt: String?
    let lowStockCount: Int
    let anomalyCount: Int
    let pendingPriceRecommendations: Int
    let pendingReorderRecommendations: Int
    let pendingStockAnomalies: Int
    let openPurchaseOrders: Int
    let activeWorkOrders: Int
    let inTransitTransfers: Int
    let purchasedUnits: Int
    let materialUnits: Int
    let finishedUnits: Int
    let recentOutcomes: [RecommendationOutcome]
}

nonisolated struct InventoryInsight: Codable, Identifiable, Hashable, Sendable {
    let inventoryId: String?
    let skuId: String
    let storeId: String
    let skuCode: String
    let description: String
    let longDescription: String?
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let costPrice: Double?
    let currentPrice: Double?
    let currentPriceValidUntil: String?
    let purchasedQty: Int
    let purchasedIncomingQty: Int
    let materialQty: Int
    let materialIncomingQty: Int
    let materialAllocatedQty: Int
    let finishedQty: Int
    let finishedAllocatedQty: Int
    let inTransitQty: Int
    let activeWorkOrderCount: Int
    let qtyOnHand: Int
    let reorderLevel: Int
    let reorderQty: Int
    let lowStock: Bool
    let anomalyFlag: Bool
    let anomalyReason: String?
    let recentSalesQty: Int
    let recentSalesRevenue: Double
    let avgDailySales: Double
    let daysOfCover: Double?
    let pendingRecommendationCount: Int
    let pendingPriceRecommendationCount: Int
    let lastUpdated: String?

    var id: String { skuId }
}

nonisolated struct ManagerRecommendation: Codable, Identifiable, Sendable {
    let id: String
    let storeId: String
    let skuId: String?
    let inventoryId: String?
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let type: RecommendationType
    let status: RecommendationStatus
    let title: String
    let rationale: String
    let confidence: Double
    let supportingMetrics: [String: JSONValue]
    let source: String
    let expectedImpact: String?
    let currentPrice: Double?
    let suggestedPrice: Double?
    let suggestedOrderQty: Int?
    let workflowAction: String?
    let analysisStatus: String
    let generatedAt: String
    let decidedAt: String?
    let appliedAt: String?
    let note: String?
}

nonisolated struct InventoryAdjustmentHistory: Codable, Identifiable, Sendable {
    let id: String
    let inventoryId: String
    let skuId: String
    let storeId: String
    let quantityDelta: Int
    let resultingQty: Int
    let reason: String
    let source: String
    let note: String?
    let createdAt: String
}

nonisolated struct ManagerAnalysisTriggerResponse: Codable, Sendable {
    let analysisStatus: String
    let multicaStatus: String
    let recommendationsCreated: Int
    let recommendationsReused: Int
    let recommendations: [ManagerRecommendation]
}

nonisolated struct SupplyChainSummary: Codable, Sendable {
    let storeId: String
    let supplierCount: Int
    let openPurchaseOrders: Int
    let activeWorkOrders: Int
    let inTransitTransfers: Int
    let purchasedUnits: Int
    let materialUnits: Int
    let finishedUnits: Int
}

nonisolated struct SupplierSummary: Codable, Identifiable, Sendable {
    let id: String
    let name: String
    let contactName: String?
    let email: String?
    let phone: String?
    let leadTimeDays: Int
    let currency: String
    let notes: String?
    let isActive: Bool
}

nonisolated struct StageInventoryPosition: Codable, Identifiable, Sendable {
    let id: String
    let storeId: String
    let skuId: String
    let skuCode: String
    let description: String
    let inventoryType: InventoryType
    let sourcingStrategy: SourcingStrategy
    let supplierName: String?
    let quantityOnHand: Int
    let incomingQuantity: Int
    let allocatedQuantity: Int
    let availableQuantity: Int
}

nonisolated struct PurchaseOrderLine: Codable, Identifiable, Sendable {
    let lineId: String
    let skuId: String
    let skuCode: String
    let description: String
    let stageInventoryType: InventoryType
    let quantity: Int
    let unitCost: Double
    let receivedQuantity: Int
    let openQuantity: Int
    let note: String?

    var id: String { lineId }
}

nonisolated struct PurchaseOrderSummary: Codable, Identifiable, Sendable {
    let id: String
    let supplierId: String
    let supplierName: String?
    let status: String
    let lines: [PurchaseOrderLine]
    let totalQuantity: Int
    let totalCost: Double
    let expectedDeliveryDate: String?
    let note: String?
    let recommendationId: String?
}

nonisolated struct WorkOrderComponent: Codable, Identifiable, Sendable {
    let skuId: String
    let skuCode: String
    let description: String
    let quantityRequired: Int
    let note: String?

    var id: String { "\(skuId)-\(quantityRequired)" }
}

nonisolated struct WorkOrderSummary: Codable, Identifiable, Sendable {
    let id: String
    let finishedSkuId: String
    let finishedSkuCode: String
    let finishedDescription: String
    let workOrderType: String
    let status: String
    let targetQuantity: Int
    let completedQuantity: Int
    let components: [WorkOrderComponent]
    let dueDate: String?
    let note: String?
    let recommendationId: String?
}

nonisolated struct StockTransferSummary: Codable, Identifiable, Sendable {
    let id: String
    let skuId: String
    let skuCode: String
    let description: String
    let quantity: Int
    let fromInventoryType: InventoryType
    let toInventoryType: InventoryType
    let status: String
    let note: String?
    let recommendationId: String?
    let dispatchedAt: String?
    let receivedAt: String?
}

nonisolated struct BOMRecipeComponent: Codable, Identifiable, Sendable {
    let skuId: String
    let skuCode: String
    let description: String
    let quantityRequired: Int
    let note: String?

    var id: String { "\(skuId)-\(quantityRequired)-\(note ?? "")" }
}

nonisolated struct BOMRecipeSummary: Codable, Identifiable, Sendable {
    let id: String
    let storeId: String
    let finishedSkuId: String
    let finishedSkuCode: String
    let finishedDescription: String
    let name: String
    let yieldQuantity: Int
    let components: [BOMRecipeComponent]
    let notes: String?
}
